libbeat/monitoring/inputmon/input.go

```diff
@@ -19,23 +19,123 @@ package inputmon
 
 import (
 	"encoding/json"
+	"errors"
+	"fmt"
 	"strings"
+	"sync"
 
 	"github.com/gofrs/uuid/v5"
 
 	"github.com/elastic/elastic-agent-libs/logp"
 	"github.com/elastic/elastic-agent-libs/monitoring"
 )
 
-// NewInputRegistry returns a new monitoring.Registry for metrics related to
-// an input instance. The returned registry will be initialized with a static
-// string values for the input and id. When the input stops it should invoke
-// the returned cancel function to unregister the metrics. For testing purposes
-// an optional monitoring.Registry may be provided as an alternative to using
-// the global 'dataset' monitoring namespace. The inputType and id must be
-// non-empty for the metrics to be published to the global 'dataset' monitoring
-// namespace.
-func NewInputRegistry(inputType, id string, optionalParent *monitoring.Registry) (reg *monitoring.Registry, cancel func()) {
+type inputRegistry struct {
+	mu         sync.RWMutex
+	registries map[string]*monitoring.Registry
+}
+
+var registeredInputs = inputRegistry{
+	registries: make(map[string]*monitoring.Registry),
+}
+
+// RegisterMetrics adds reg to the collection of registries to be returned by
+// the `/inputs/` endpoint. The registry must have at least a `id` and `input`
+// string variables, otherwise the registry is rejected and an error is
+// returned.
+// If an id/inputType registry has benn already registered, it'll be overridden.
+// When the input finishes, it should call UnregisterMetrics to
+// release the associated resources.
+func RegisterMetrics(id string, reg *monitoring.Registry) error {
```

Comment:
Is there a reason to require each input to handle this manually? This seems like another thing the pipeline could take care of transparently via the `Context` API when the input first requests the registry (or even earlier, unless there's a reason we specifically want to prevent some inputs from showing up in metrics)

Reply:
> Is there a reason to require each input to handle this manually? 

It depends :)

As the inputs still need to make a call to properly add their type and id to the registry, so we cannot avoid that for now, I just thought they could then "decide" to publish metrics or not. It basically would avoid the case an input does not published metrics to have the per-input pipeline metrics published.

Reply:
If the input doesn't publish its own metrics (and thus doesn't explicitly provide a type and id), we can just use the default values the pipeline already knows about, right? I thought we only needed the inputs to specify those at all for backwards compatibility with inputs that report different names / ids than the canonical ones.